Bug 1186100 - Making 'ls.lock-immediately' lock immediately

cr opened the pull request:
This is a proposed fix for [bug 1186100](http://bugzil.la/1186100) where 'lockscreen.lock-immediately'
as sent by _FindMyDevice_ doesn't lock the device immediately with a pass code
when a pass code timeout has been set.

On unlock, `LockScreen.checkPassCodeTimeout()` only asks for the pass code
when either `.fetchLockedInterval()` or `.fetchUnlockedInterval()` exceed
the pass code timeout.

Setting both `_lastLockedTimeStamp` and `_lastUnlockedTimeStamp` to
something deep in the past will enforce pass code entry on unlock and
avoid a potential race condition.

Commit:
Making 'ls.lock-immediately' lock immediately

This fixes a bug where 'lockscreen.lock-immediately' as sent by
FindMyDevice doesn't lock the device immediately with a passcode
when a pass code timeout has been set.

On unlock, LockScreen.checkPassCodeTimeout() only asks for the pass code
when either .fetchLockedInterval() or .fetchUnlockedInterval() exceed
the pass code timeout.

Setting both _lastLockedTimeStamp and _lastUnlockedTimeStamp to
something deep in the past will enforce pass code entry on unlock and
avoid a potential race condition.

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8874b731f2bd0fadc775842c4e1c986b392cfe50

Commit:
Moving 'ls.lock-immediately' fix to lockscreen.js

Moved timestamp reset call from LockScreenWindowManager.'s
'lockscreen.lock-immediately' handler to LockScreen's, fully
encapsulating the fix for bug 1186100 in lockscreen.js.

Added inline comments on bugfix rationale.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=54816bd906e8bf9851c437ed8dbf1d24003983d2

Commit:
Add and fix lock-immediately tests

This extends lockscreen_test.js with two new tests that check whether
resetting timeouts will trigger pass code check, and whether lock-immediately
triggers timeout reset.

This also fixes the old lock-immediately test which didn't trigger the
observer callback and effectively tested whether a locked lockscreen would be
locked, which resulted in a broken test that never fails.

Commit:
Port settings observer to SettingsListener.observe()

The only reason for this is consistency with the rest of the file. Also
this makes writing tests slightly more pleasing.

Commit:
Move resetTimeoutForcibly close to its relatives

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2b533e66a2a148a08faf4038d49156945c41d2f5

lissyx commented on apps/system/lockscreen/js/lockscreen.js:
```diff
@@ -434,10 +434,14 @@
-    navigator.mozSettings.addObserver('lockscreen.lock-immediately',
-      (function(event) {
-      if (event.settingValue === true) {
+    window.SettingsListener.observe('lockscreen.lock-immediately',
```
Please make sure that this is executed only once. This codes lives in the system app, so adding mozSettings observers continuously will leak.

cr replied:
LockScreen registers its settings observers as part of LockScreen.init() which is only ever called once when System starts here:

http://mxr.mozilla.org/gaia/source/apps/system/lockscreen/js/lockscreen_bootstrap.js#27

Do you think this will require deeper verification?

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ca266291a40dc1ee4b31fe537a845ec7204242da

Commit:
Adapted .checkPassCodeTimeout() argument

Commit:
Request lock from state manager instead of direct .lock

With the new lockscreen state manager in place, the 'lock-immediately'
must send 'lockscreen-request-lock' event instead of using the .lock
method directly. Else, state manager will lose track of lock state.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7e82fcb816aaf908a319ed9c182610e8078dc346

snowmantw commented on apps/system/lockscreen/js/lockscreen.js:
```diff
@@ -351,9 +351,13 @@
       // bug 992277. We don't need to use SettingsListener because
```
Since now you use the SettingsListener, you may fix the comment, too.

cr replied:
I don't know who wrote this comment, but it is still valid and should stay. The point it makes is that we should not need to misuse a settings listener for this event-style signalling. I didn't touch that part of the code since I wasn't involved in that design.
